python/dgl/heterograph.py

```diff
@@ -0,0 +1,1214 @@
+            Reduce function on the node. The function should be
+            a :mod:`Node UDF <dgl.udf>`.
+        apply_node_func : callable, optional
+            Apply function on the nodes. The function should be
+            a :mod:`Node UDF <dgl.udf>`.
+
+        Notes
+        -----
+        An error is raised if the graph contains multiple node types *or*
+        multiple edge types.
+        """
+        pass
+
+    # TODO should we support this?
+    def prop_nodes(self,
+                   nodes_generator,
+                   message_func="default",
+                   reduce_func="default",
+                   apply_node_func="default"):
+        """Node propagation in heterogeneous graph is not supported.
+        """
+        raise NotImplementedError('not supported')
+
+    # TODO should we support this?
+    def prop_edges(self,
+                   edges_generator,
+                   message_func="default",
+                   reduce_func="default",
+                   apply_node_func="default"):
+        """Edge propagation in heterogeneous graph is not supported.
+        """
+        raise NotImplementedError('not supported')
+
+    def subgraph(self, nodes):
+        """Return the subgraph induced on given nodes.
+
+        Parameters
+        ----------
+        nodes : dict[str, list or iterable]
+            A dictionary of node types to node ID array to construct
+            subgraph.
+            All nodes must exist in the graph.
+
+        Returns
+        -------
+        G : DGLHeteroSubGraph
+            The subgraph.
+            The nodes are relabeled so that node `i` of type `t` in the
+            subgraph is mapped to the ``nodes[i]`` of type `t` in the
+            original graph.
+            The edges are also relabeled.
+            One can retrieve the mapping from subgraph node/edge ID to parent
+            node/edge ID via `parent_nid` and `parent_eid` properties of the
+            subgraph.
+        """
+        pass
+
+    def subgraphs(self, nodes):
+        """Return a list of subgraphs, each induced in the corresponding given
+        nodes in the list.
+
+        Equivalent to
+        ``[self.subgraph(nodes_list) for nodes_list in nodes]``
+
+        Parameters
+        ----------
+        nodes : a list of dict[str, list or iterable]
+            A list of type-ID dictionaries to construct corresponding
+            subgraphs.  The dictionaries are of the same form as
+            :func:`subgraph`.
+            All nodes in all the list items must exist in the graph.
+
+        Returns
+        -------
+        G : A list of DGLHeteroSubGraph
+            The subgraphs.
+        """
+        pass
+
+    def edge_subgraph(self, edges):
+        """Return the subgraph induced on given edges.
+
+        Parameters
+        ----------
+        edges : dict[etype, list or iterable]
+            A dictionary of edge types to edge ID array to construct
+            subgraph.
```

Comment:
edge types are strings or tuples of strings?

Reply:
Tuple of strings (see L946)